pkg/tests/ossm/smoke_test.go

```diff
@@ -19,50 +19,195 @@ import (
+
+			t.LogStep("Check whether sidecar is injected in all bookinfo pods")
+			assertSidecarInjectedInAllBookinfoPods(t, ns)
+
+			t.LogStep("Check if bookinfo productpage is running through the Proxy")
+			retry.UntilSuccess(t, func(t test.TestHelper) {
+				oc.Exec(t,
+					pod.MatchingSelector("app=sleep", ns), "sleep",
+					"curl -sI http://productpage:9080",
+					assert.OutputContains(
+						"HTTP/1.1 200 OK",
+						"ProductPage returns 200 OK",
+						"ProductPage didn't return 200 OK"),
+					assert.OutputContains(
+						"server: istio-envoy",
+						"HTTP header 'server: istio-envoy' is present in the response",
+						"HTTP header 'server: istio-envoy' is missing from the response"),
+					assert.OutputContains(
+						"x-envoy-decorator-operation",
+						"HTTP header 'x-envoy-decorator-operation' is present in the response",
+						"HTTP header 'x-envoy-decorator-operation' is missing from the response"))
+			})
+
+			t.LogStep("verify proxy startup time. Expected to be less than 10 seconds")
+			t.Log("Jira related: https://issues.redhat.com/browse/OSSM-3586")
+			t.Log("From proxy yaml, verify the time between status.containerStatuses.state.running.startedAt and status.conditions[type=Ready].lastTransitionTime")
+			ratingPod := pod.MatchingSelector("app=ratings", ns)(t, oc.DefaultOC)
+			ratingYaml := oc.GetYaml(t, ns, "pod", ratingPod.Name)
+			t.Log("Validate from YAML, the proxy startup time is less than 10 seconds for ratings pod")
+			validateStartUpProxyTime(t, ratingYaml)
 		})
 
-		DeployControlPlane(t)
-
-		t.LogStep("Install bookinfo pods with sidecar and sleep pod without sidecar")
-		app.InstallAndWaitReady(t, app.Bookinfo(ns), app.SleepNoSidecar(ns))
-
-		t.LogStep("Check whether sidecar is injected in all bookinfo pods")
-		assertSidecarInjectedInAllBookinfoPods(t, ns)
-
-		t.LogStep("Check if bookinfo productpage is running through the Proxy")
-		oc.Exec(t,
-			pod.MatchingSelector("app=sleep", ns), "sleep",
-			"curl -sI http://productpage:9080",
-			assert.OutputContains(
-				"HTTP/1.1 200 OK",
-				"ProductPage returns 200 OK",
-				"ProductPage didn't return 200 OK"),
-			assert.OutputContains(
-				"server: istio-envoy",
-				"HTTP header 'server: istio-envoy' is present in the response",
-				"HTTP header 'server: istio-envoy' is missing from the response"),
-			assert.OutputContains(
-				"x-envoy-decorator-operation",
-				"HTTP header 'x-envoy-decorator-operation' is present in the response",
-				"HTTP header 'x-envoy-decorator-operation' is missing from the response"))
+		t.NewSubTest(fmt.Sprintf("upgrade %s to %s", fromVersion, toVersion)).Run(func(t TestHelper) {
+			t.Logf("This test checks whether SMCP becomes ready after it's upgraded from %s to %s and bookinfo is still working after the upgrade", fromVersion, toVersion)
+
+			t.Cleanup(func() {
+				app.Uninstall(t, app.Bookinfo(ns), app.SleepNoSidecar(ns))
+			})
+
+			t.LogStepf("Upgrade SMCP from %s to %s", fromVersion, toVersion)
```

Comment:
I wonder: Given that we must support running control plane version x.y with proxy version x.(y-1), shouldn't we (also) run all the entire test suite with this scenario (using the two different versions). This is the only way we can be sure that the behavior is correct. It's critical that everything works correctly even if there's a version difference between the control plane and the data plane. Or do we already test this?

Reply:
We are doing that test because we made the upgrade, verify that the traffic is working and them restart the pods to get the new proxy and verify agains the traffic. So this scenario is covered:
```
t.LogStep("Check if bookinfo productpage is running through the Proxy after the upgrade")
			retry.UntilSuccess(t, func(t test.TestHelper) {
				oc.Exec(t,
					pod.MatchingSelector("app=sleep", ns), "sleep",
					"curl -sI http://productpage:9080",
					assert.OutputContains(
						"HTTP/1.1 200 OK",
						"ProductPage returns 200 OK",
						"ProductPage didn't return 200 OK"),
					assert.OutputContains(
						"server: istio-envoy",
						"HTTP header 'server: istio-envoy' is present in the response",
						"HTTP header 'server: istio-envoy' is missing from the response"),
					assert.OutputContains(
						"x-envoy-decorator-operation",
						"HTTP header 'x-envoy-decorator-operation' is present in the response",
						"HTTP header 'x-envoy-decorator-operation' is missing from the response"))
			})

			t.LogStep("Delete Bookinfo pods to validate proxy is still working after recreation and upgrade")
			oc.RestartAllPodsAndWaitReady(t, ns)
			retry.UntilSuccess(t, func(t test.TestHelper) {
				oc.Exec(t,
					pod.MatchingSelector("app=sleep", ns), "sleep",
					"curl -sI http://productpage:9080",
					assert.OutputContains(
						"HTTP/1.1 200 OK",
						"ProductPage returns 200 OK",
						"ProductPage didn't return 200 OK"),
					assert.OutputContains(
						"server: istio-envoy",
						"HTTP header 'server: istio-envoy' is present in the response",
						"HTTP header 'server: istio-envoy' is missing from the response"),
					assert.OutputContains(
						"x-envoy-decorator-operation",
						"HTTP header 'x-envoy-decorator-operation' is present in the response",
						"HTTP header 'x-envoy-decorator-operation' is missing from the response"))
			})
```

Reply:
Yes, but this only checks the very basic use-case (whether the traffic flows through the proxy). It doesn't check everything else that can be configured (e.g. traffic splitting, circuit breaking, mTLS, etc.). How do we know those features work properly when the control plane version is not the same as the data plane version?

Reply:
But to make this happens we will increase to the double the execution time, yes will be more coverage and verification. But maybe this can be a feature to run in night builds where time does not care at the same level than in daily builds. wdyt?

Reply:
Yes, this should be a separate test run, just like the upgrade, interop, and other test types. Ideally, we would add two additional test runs: with control plane version being 1 more than the data plane version, and with the control plane version being 1 less than the data plane version (to test downgrades).